File: src/languagesMap/js/TreeNode/implementation.js
class TreeNode {
  constructor(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
  }
}

function arrayToBinaryTree(arrayNode) {
  if (!arrayNode.length || arrayNode[0] === null) {
    return null
  }
  const root = new TreeNode(arrayNode[0])
  const nodes = [root]
  for (let i = 0; i < arrayNode.length && nodes.length; i++) {
    const node = nodes.shift()
    if (2*i+1 < arrayNode.length && arrayNode[2*i+1] !== null) {
      const left = new TreeNode(arrayNode[2*i+1])
      node.left = left
      nodes.push(left)
    }
    if (2*i+2 < arrayNode.length && arrayNode[2*i+2] !== null) {
      const right = new TreeNode(arrayNode[2*i+2])
      node.right = right
      nodes.push(right)
    }
  }
  return root
}

function binaryTreeToArray(root) {
  if (!root) return [];
  const arr = []
  const nodes = [root]
  while (nodes.length) {
    const node = nodes.shift()
    if (!node) {
      arr.push(null)
      continue
    } else {
      arr.push(node.val)
    }
    nodes.push(node.left)
    nodes.push(node.right)
  }
  while (arr[arr.length - 1] === null) {
    arr.pop()
  }
  return arr
}

File: src/languagesMap/py/TreeNode/implementation.py
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right
def arrayToBinaryTree(arrayNode):
  if len(arrayNode) == 0 or arrayNode[0] == None:
    return None
  
  root = TreeNode(arrayNode[0])
  nodes = [root]
  i = 0
  while i < len(arrayNode) and len(nodes) > 0:
    node = nodes.pop(0)
    if 2*i+1 < len(arrayNode) and arrayNode[2*i+1] != None:
      left = TreeNode(arrayNode[2*i+1])
      node.left = left
      nodes.append(left)
  
    if 2*i+2 < len(arrayNode) and arrayNode[2*i+2] != None:
      right = TreeNode(arrayNode[2*i+2])
      node.right = right
      nodes.append(right)
  
    i += 1
  
  return root

def binaryTreeToArray(root):
  if root == None:
    return []
  
  arr = []
  nodes = [root]
  while len(nodes) > 0:
    node = nodes.pop(0)
    if node == None:
      arr.append(None)
      continue  
    else:
      arr.append(node.val)
    nodes.append(node.left)
    nodes.append(node.right)
  while arr[-1] == None:
    arr.pop()
  return arr
